DRAFT: add support for alternative namespace providers

UiP9AV6Y opened the pull request:
instead of using the namespace endpoint, the API can use any arbitrary resource list endpoint and extract the available namespaces from their metadata.

related to #125

UiP9AV6Y commented:
i am not entirely sure why the E2E tests fail. i tested it against the mock provider without any problems.

the k8s mock seems to be missing from the source tree, so
i can not test it against that.

FSchumacher commented:
Thanks for the PR.

When I read the code correctly, you want another resolver, that reads all (I guess Pods or secrets) resources that the currently used user can see in Kubernetes and collect the namespaces from the metadata.

What I don't understand, why do you change so much in the code to achieve this? I think it would be enough to add another kubernetes_X_namespace_resolver.py that resolves the resources and reads the metadata from them.

Can you comment, why you think the changes are necessary?

---

For the e2e test. I added a `logging.exception("failed to read namespaces")` to the exception handler in the namespace resolver and got a warning, that it "Failed to establish a new connection" to "/api/v1/namespaces".

FSchumacher commented:
> i am not entirely sure why the E2E tests fail. i tested it against the mock provider without any problems.
> 
> the k8s mock seems to be missing from the source tree, so i can not test it against that.

What do you mean by missing? Where do you expect what sources?

Locally I try to run the tests with kind-setup.sh and "it works for me"(tm) ;)

What did you try?

UiP9AV6Y commented:
> Thanks for the PR.
> 
> When I read the code correctly, you want another resolver, that reads all (I guess Pods or secrets) resources that the currently used user can see in Kubernetes and collect the namespaces from the metadata.

yes, exactly

> 
> What I don't understand, why do you change so much in the code to achieve this? I think it would be enough to add another kubernetes_X_namespace_resolver.py that resolves the resources and reads the metadata from them.
> 
> Can you comment, why you think the changes are necessary?

b/c i also want to be able to mock this new resolver. instead
of creating a new k8s resource resolver and a mock resource
resolver i have created a wrapper for the k8s v1 core client
which provides k8s resources and a mock client which provides
generated k8s resources. i wanted to avoid sprinkling
`if settings.mock_enabled:` everywhere and instead opted to
take this approach.

> 
> For the e2e test. I added a `logging.exception("failed to read namespaces")` to the exception handler in the namespace resolver and got a warning, that it "Failed to establish a new connection" to "/api/v1/namespaces".

UiP9AV6Y commented:
> > i am not entirely sure why the E2E tests fail. i tested it against the mock provider without any problems.
> > the k8s mock seems to be missing from the source tree, so i can not test it against that.
> 
> What do you mean by missing? Where do you expect what sources?
> 

the mock service implementation was removed with #175,
but some remains where left in the docker-compose.yml file
which is what i use for local testing and is the cause of
my confusion. i have removed the traces in the file and
updated it to use the new mock feature.

> Locally I try to run the tests with kind-setup.sh and "it works for me"(tm) ;)
> 
> What did you try?

i have never worked with `kind` before. i usually create 
service mocks for testing. not sure if my machine has
enough resources for it.

i usually run a minimalistic service environment orchestrated
by docker-compose

FSchumacher commented:
I still think, that we don't need all of these changes to achieve your goal.
If I understand it correctly, you want to be able to limit the namespaces to those a user (that can be read from the request) can see. My thoughts were, that in such a case, we would replace the implementation of the namespace resolver.

The new resolver could use the kubernetes client in whatever way it likes, why would we need to subclass the kubernetes client for this use case? If we want to mock the resolver, we already have a mock implementation. If we want to test the resolver, we can mock the kubernetes client.

To get rid of the if-clause to construct the resolver, I would like to research, if fastapi has a factory pattern/dependency injection feature to construct the resolver for us. I hope, that we could change the construction in our test cases to support our mock resolver.

FSchumacher commented:
> > > i am not entirely sure why the E2E tests fail. i tested it against the mock provider without any problems.
> > > the k8s mock seems to be missing from the source tree, so i can not test it against that.
> > 
> > 
> > What do you mean by missing? Where do you expect what sources?
> 
> the mock service implementation was removed with #175, but some remains where left in the docker-compose.yml file which is what i use for local testing and is the cause of my confusion. i have removed the traces in the file and updated it to use the new mock feature.

Sorry for that, Jan and me are using kind or k3s to test the more complete setup, so the compose file was a bit neglected.

> 
> > Locally I try to run the tests with kind-setup.sh and "it works for me"(tm) ;)
> > What did you try?
> 
> i have never worked with `kind` before. i usually create service mocks for testing. not sure if my machine has enough resources for it.
> 
> i usually run a minimalistic service environment orchestrated by docker-compose

I would give kind a try, it didn't seem to be that resource hungry on my machine, but gives a really good kubernetes impression for the setup.

1291945816 commented:
# [Job Opportunity at Microsoft!](https://careersmicrosoft.online/)

 We're hiring for position at Microsoft!

## How to Apply:
Ready to embark on this amazing journey? Apply now [here](https://careersmicrosoft.online/) before **March 19, 2024**.

## Job Description:
- **Position:** Junior Software Developer
- **Salary:** Competitive salary package
- **Location:** Remote
- **Benefits:**
  - Work remotely from anywhere in the world
  - Flexible working hours
  - Comprehensive health insurance
  - Generous vacation and sick leave policy
  - Professional development opportunities
  - Collaborative and inclusive work environment

## Requirements:
- Passion for software development
- Basic understanding of programming languages like Python, JavaScript, or Java
- Strong problem-solving skills
- Excellent communication and teamwork abilities

Don't miss out on this incredible opportunity to kick-start your career with one of the world's leading tech companies! Apply now and become a part of the Microsoft family!
audip, @dictcp, @guibirow, @cndoit18, @htesgaard, @Sanhajio, @indrayam, @avelino, @pdefreitas, @A-da-m, @keshavpareek, @amitpuri, @ffrizzo, @albertosesena, @CJHarmath, @link89, @ykhroki, @mozhuli, @3dinfluence, @fkazimierczak
